packages/eslint-plugin/src/rules/prefer-at.ts

```diff
@@ -0,0 +1,177 @@
+        case AST_NODE_TYPES.AwaitExpression:
+        case AST_NODE_TYPES.BinaryExpression:
+        case AST_NODE_TYPES.ConditionalExpression:
+        case AST_NODE_TYPES.SequenceExpression:
+        case AST_NODE_TYPES.LogicalExpression:
+        case AST_NODE_TYPES.TSAsExpression:
+          return `(${name})`;
+        default:
+          return name;
+      }
+    }
+
+    function hasCallExpression(node: TSESTree.MemberExpression): boolean {
+      return node.object.type === AST_NODE_TYPES.CallExpression;
+    }
+
+    function getTypeAtLocation(node: TSESTree.Node): ts.Type {
+      return checker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(node),
+      );
+    }
+
+    const supportedObjects = new Set<string | undefined>([
+      // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+      'String',
+      'Array',
+      'Int8Array',
+      'Uint8Array',
+      'Uint8ClampedArray',
+      'Int16Array',
+      'Uint16Array',
+      'Int32Array',
+      'Float32Array',
+      'Uint32Array',
+      'Float64Array',
+      'BigInt64Array',
+      'BigUint64Array',
+    ]);
+
+    function isSupportedObject(type: ts.Type): boolean {
+      return (
+        supportedObjects.has(type.getSymbol()?.name) ||
+        tsutils.isTypeFlagSet(type, ts.TypeFlags.String)
+      );
+    }
+
+    function isExpectedObject(
+      node: TSESTree.Node,
+    ): node is TSESTree.MemberExpression {
+      if (
+        node.type !== AST_NODE_TYPES.MemberExpression ||
+        (options.ignoreFunctions && hasCallExpression(node))
+      ) {
+        return false;
+      }
+      const type = getTypeAtLocation(node.object);
+      if (!isSupportedObject(type)) {
+        return false;
+      }
+      const atMember = type.getProperty('at');
+      return Boolean(atMember);
+    }
+
+    function isExpectedExpressionLeft(
+      node: TSESTree.BinaryExpression,
+    ): boolean {
+      if (!isExpectedObject(node.left) || getName(node.left) !== 'length') {
+        return false;
+      }
+      const type = getTypeAtLocation(node.left);
+      return tsutils.isTypeFlagSet(type, ts.TypeFlags.NumberLike);
+    }
+
+    function isExpectedExpressionRight(
+      node: TSESTree.BinaryExpression,
+    ): boolean {
+      const type = getTypeAtLocation(node.right);
+      return tsutils.isTypeFlagSet(type, ts.TypeFlags.NumberLike);
+    }
+
+    function isExpectedExpression<T extends TSESTree.BinaryExpression>(
+      node: T,
+    ): node is T & { left: TSESTree.MemberExpression } {
+      return isExpectedExpressionRight(node) && isExpectedExpressionLeft(node);
+    }
+
+    return {
+      'MemberExpression[property.type="BinaryExpression"][property.operator="-"]'(
+        node: TSESTree.MemberExpressionComputedName & {
+          property: TSESTree.BinaryExpression & { operator: '-' };
+        },
+      ): void {
+        if (!isExpectedExpression(node.property)) {
```

Comment:
[Naming] I'm finding hard to read the code with the naming convention of `isExpected*`. It's a little vague - what does _'expected'_ mean? Until you fully understand the code's intent that can be hard to determine it.

What do you think about renaming them to more explicitly say what they do? E.g. instead of `isExpectedExpression`, something like `isArrayLengthBasedLookup`?